zerver/views/message_summary.py

```diff
@@ -0,0 +1,131 @@
+    # generate summaries for new messages in a single pass.
+    # TODO: Come up with a plan to store these summaries in the database.
+
+    messages_response = get_messages_backend(
+        request,
+        user_profile,
+        narrow=narrow,
+        anchor_val="newest",
+        num_before=MAX_MESSAGES_SUMMARIZED,
+        client_gravatar=False,
+        apply_markdown=False,
+    )
+    zulip_messages = json.loads(messages_response.content).get("messages", [])
+    if len(zulip_messages) == 0:
+        return json_success(request, {"summary": "No messages in conversation to summarize"})
+
+    # XXX: Translate input and output text to English?
+    model = settings.TOPIC_SUMMARIZATION_MODEL
+    litellm_params: dict[str, Any] = {}
+    if model.startswith("huggingface"):
+        assert settings.HUGGINGFACE_API_KEY is not None
+        litellm_params["api_key"] = settings.HUGGINGFACE_API_KEY
+    else:
+        assert model.startswith("bedrock")
+        litellm_params["aws_access_key_id"] = settings.AWS_ACCESS_KEY_ID
+        litellm_params["aws_secret_access_key"] = settings.AWS_SECRET_ACCESS_KEY
+        litellm_params["aws_region_name"] = settings.AWS_REGION_NAME
+
+    conversation_length = len(zulip_messages)
+    max_summary_length = get_max_summary_length(conversation_length)
+    intro = "The following is a chat conversation in the Zulip team chat app."
+    topic: str | None = None
+    channel: str | None = None
+    if narrow and len(narrow) == 2:
+        for term in narrow:
+            assert not term.negated
+            if term.operator == "channel":
+                channel = term.operand
+            if term.operator == "topic":
+                topic = term.operand
+    if channel:
+        intro += f" channel: {channel}"
+    if topic:
+        intro += f", topic: {topic}"
+
+    formatted_conversation = format_conversation(zulip_messages)
+    prompt = (
+        f"Succinctly summarize this conversation based only on the information provided, "
+        f"in up to {max_summary_length} sentences, for someone who is familiar with the context. "
+        f"Mention key conclusions and actions, if any. Refer to specific people as appropriate. "
+        f"Don't use an intro phrase."
+    )
+    messages = [
+        make_message(intro, "system"),
+        make_message(formatted_conversation),
+        make_message(prompt),
+    ]
+
+    # Token counter is recommended by LiteLLM but mypy says it's not explicitly exported.
+    # https://docs.litellm.ai/docs/completion/token_usage#3-token_counter
+    input_tokens = litellm.token_counter(model=model, messages=messages)  # type: ignore[attr-defined] # Explained above
+    response = litellm.completion(
+        model=model,
+        messages=messages,
+        **litellm_params,
+    )
+    output_tokens = response["usage"]["completion_tokens"]
+
+    credits_used = (output_tokens * OUTPUT_COST_PER_TOKEN) + (input_tokens * INPUT_COST_PER_TOKEN)
```

Comment:
So most of the models that we are considering charge the same for input and output tokens. I would kind of expect the data in `completion` just has the number of inputs tokens used in another field of `"usage"`; does it not? 

https://docs.litellm.ai/docs/completion/token_usage

Reply:
I guess it depends whether we pay for the request to ask how many tokens there are :). If so, input tokens are a large portion of total token volume, so paying for that request likely doesn't make any sense.

(We can use other methods to ensure one never goes over, but for v1, I think it's fine to allow a final request that pushes you just over the limit; it's hard to make an individual request big enough).

Reply:
I guess most correct would be:

- Before querying the model, check whether we've enough tokens left using the estimated token count.
- Then increase the `LoggingCountStat` using the estimated token count. (These first two steps should be a short database transaction that locks the `LoggingCountStat` row).
- Then query the model.
- Then adjust the `LoggingCountStat` by `(actual - estimated)`, being careful to avoid doing this to the next day if the query happened milliseconds before midnight; changing the `LoggingCountStat` we added the estimate to.

That way, you can't easily get extra tokens by sending 25 requests all at once when you're just below the limit.

Reply:
I added that as a TODO to implement once we have more clarity on user plans.